### PR TITLE
Feature/h3_integration

### PR DESCRIPTION
**Pull request by edurdevic**
Enables support for DBR 11.2 with native Databricks integration of H3

### CODE REVIEW

File: docs/source/usage/automatic-sql-registration.rst [hunk 1]
```diff
@@ -14,7 +14,7 @@ Pre-requisites
 **************
 
 In order to use Mosaic, you must have access to a Databricks cluster running
-Databricks Runtime 10.0 or later. If you have cluster creation permissions in your Databricks
+Databricks Runtime 11.2 or later. If you have cluster creation permissions in your Databricks
```
Comment (milos-colic) [inline docs/source/usage/automatic-sql-registration.rst:17]: @edurdevic will it still work before 11.2 with the fallback?
If yes, I would say 10.0 is prerequisite, 11.2 is recommended.
Maybe we need to reword this sentence.

File: docs/source/usage/installation.rst [hunk 2]
```diff
@@ -5,7 +5,7 @@ Installation guide
 Supported platforms
 ###################
 In order to use Mosaic, you must have access to a Databricks cluster running
-Databricks Runtime 10.0 or later. If you have cluster creation permissions in your Databricks
+Databricks Runtime 11.2 or later. If you have cluster creation permissions in your Databricks
```
Comment (milos-colic) [inline docs/source/usage/installation.rst:8]: @edurdevic same as for automatic-sql-registration.rst

File: notebooks/examples/scala/QuickstartNotebook.scala [hunk 3]
```diff
@@ -17,7 +17,7 @@ print(s"The raw data is stored in $raw_path")
 
 // MAGIC %md
 // MAGIC ## Enable Mosaic in the notebook
-// MAGIC Mosaic requires Databricks Runtime (DBR) version 10.0 or higher. </br>
+// MAGIC Mosaic requires Databricks Runtime (DBR) version 11.2 or higher. </br>
```
Comment (milos-colic) [inline notebooks/examples/scala/QuickstartNotebook.scala:20]: @edurdevic same as for automatic-sql-registration

File: python/README.md [hunk 4]
```diff
@@ -1,107 +0,0 @@
-# Mosaic by Databricks Labs
```
Comment (milos-colic) [inline python/README.md:1]: @edurdevic why is this file deleted?

File: src/main/scala/com/databricks/labs/mosaic/core/geometry/point/MosaicPoint.scala [hunk 5]
```diff
@@ -3,7 +3,7 @@ package com.databricks.labs.mosaic.core.geometry.point
 import com.databricks.labs.mosaic.core.geometry.MosaicGeometry
 import com.databricks.labs.mosaic.core.MosaicCoreException
 import com.databricks.labs.mosaic.core.geometry.linestring.MosaicLineString
-import com.uber.h3core.util.GeoCoord
+import com.databricks.labs.mosaic.core.types.model.Coordinates
```
Comment (milos-colic) [inline src/main/scala/com/databricks/labs/mosaic/core/geometry/point/MosaicPoint.scala:6]: @edurdevic I like abstraction from external GeoCoord type coming from H3.
We may need to rename some of the methods like getGeoCoords and similar. 
Not critical at all at the moment. Just something to keep in mind.

File: src/main/scala/com/databricks/labs/mosaic/core/index/H3IndexSystem.scala [hunk 6]
```diff
@@ -85,7 +86,10 @@ object H3IndexSystem extends IndexSystem with Serializable {
     override def indexToGeometry(index: Long, geometryAPI: GeometryAPI): MosaicGeometry = {
         val boundary = h3.h3ToGeoBoundary(index).asScala
         val extended = boundary ++ List(boundary.head)
-        geometryAPI.geometry(extended.map(geometryAPI.fromGeoCoord), POLYGON)
+        geometryAPI.geometry(
+          extended.map(p => geometryAPI.fromGeoCoord(Coordinates(p.lat, p.lng))),
```
Comment (milos-colic) [inline src/main/scala/com/databricks/labs/mosaic/core/index/H3IndexSystem.scala:90]: @edurdevic I would add a Coordinates(GeoCoord) constructor so that we could keep simpler map expression.
Very minor, and I can do it as a part of some other PR.
I am very biased towards 1-liners :)

File: src/main/scala/com/databricks/labs/mosaic/functions/MosaicContext.scala [hunk 7]
```diff
@@ -654,19 +672,19 @@ object MosaicContext {
 
     def build(indexSystem: IndexSystem, geometryAPI: GeometryAPI): MosaicContext = {
         instance = Some(new MosaicContext(indexSystem, geometryAPI))
-        context
+        context()
     }
 
-    def context: MosaicContext =
+    def context(): MosaicContext =
```
Comment (milos-colic) [inline src/main/scala/com/databricks/labs/mosaic/functions/MosaicContext.scala:678]: @edurdevic these dont really need () no side effects other than Error throw. 
It may simplify jvm calls from spark so works for me.

File: src/test/scala/com/databricks/labs/mosaic/expressions/index/TestPointIndex.scala [hunk 8]
```diff
@@ -15,13 +15,13 @@ class TestPointIndex extends QueryTest with SharedSparkSession with PointIndexBe
...
 
-    test("Testing stArea (H3, JTS) NO_CODEGEN") { noCodegen { wktPointIndex(H3IndexSystem, JTS, 5) } }
-    test("Testing stArea (H3, ESRI) NO_CODEGEN") { noCodegen { wktPointIndex(H3IndexSystem, ESRI, 5) } }
-    test("Testing stArea (BNG, JTS) NO_CODEGEN") { noCodegen { wktPointIndex(BNGIndexSystem, JTS, 5) } }
-    test("Testing stArea (BNG, ESRI) NO_CODEGEN") { noCodegen { wktPointIndex(BNGIndexSystem, ESRI, 5) } }
-    test("Testing stArea auxiliaryMethods (H3, JTS)") { noCodegen { auxiliaryMethods(H3IndexSystem, JTS) } }
-    test("Testing stArea auxiliaryMethods (H3, ESRI)") { noCodegen { auxiliaryMethods(H3IndexSystem, ESRI) } }
-    test("Testing stArea auxiliaryMethods (BNG, JTS)") { noCodegen { auxiliaryMethods(BNGIndexSystem, JTS) } }
-    test("Testing stArea auxiliaryMethods (BNG, ESRI)") { noCodegen { auxiliaryMethods(BNGIndexSystem, ESRI) } }
+    test("Testing PointIndexGeom (H3, JTS) NO_CODEGEN") { noCodegen { wktPointIndex(H3IndexSystem, JTS, 5) } }
```
Comment (milos-colic) [inline src/test/scala/com/databricks/labs/mosaic/expressions/index/TestPointIndex.scala:18]: @edurdevic glad you caught this error in naming!

File: src/test/scala/com/databricks/sql/package.scala [hunk 9]
```diff
@@ -0,0 +1,24 @@
+package com.databricks.sql
+
+import org.apache.spark.sql.Column
+import org.apache.spark.sql.functions.lit
+
+object functions {
```
Comment (milos-colic) [inline src/test/scala/com/databricks/sql/package.scala:6]: @edurdevic can we just add here a comment explicitly stating why we added this object.
I know it is rather obvious from the dummy function definitions, but still it is better if we added that comment

Comment (milos-colic) [inline src/test/scala/com/databricks/sql/package.scala:6]: @edurdevic can we just add here a comment explicitly stating why we added this object.
I know it is rather obvious from the dummy function definitions, but still it is better if we added that comment

**Review by milos-colic — APPROVED** [review]
A couple of minor comments. Nothing blocking the merge to be honest.

File: docs/source/usage/automatic-sql-registration.rst (hunk 1 above)
Comment (edurdevic) [reply docs/source/usage/automatic-sql-registration.rst:17]: Yes, you are right! 
I wrote this comment before we decided to include the H3 JAR in Mosaic as well.

File: notebooks/examples/scala/QuickstartNotebook.scala (hunk 3 above)
Comment (edurdevic) [reply notebooks/examples/scala/QuickstartNotebook.scala:20]: Updated

File: python/README.md (hunk 4 above)
Comment (edurdevic) [reply python/README.md:1]: This was an outdated copy of the main README.

File: src/main/scala/com/databricks/labs/mosaic/core/index/H3IndexSystem.scala (hunk 6 above)
Comment (edurdevic) [reply src/main/scala/com/databricks/labs/mosaic/core/index/H3IndexSystem.scala:90]: We could, but that would introduce an H3 dependency from our core functions.
I prefer to keep all the H3 references within this file. Maybe a helper function within this file to shorten the map expression?

File: src/main/scala/com/databricks/labs/mosaic/functions/MosaicContext.scala (hunk 7 above)
Comment (edurdevic) [reply src/main/scala/com/databricks/labs/mosaic/functions/MosaicContext.scala:678]: I added those just to get rid of the warnings that IntelliJ was giving me...

File: src/test/scala/com/databricks/sql/package.scala (hunk 9 above)
Comment (edurdevic) [reply src/test/scala/com/databricks/sql/package.scala:6]: Good point!

Comment (edurdevic) [reply src/test/scala/com/databricks/sql/package.scala:6]: Done

**Comment by edurdevic** [comment]
@mickvanhulst-TomTom This PR will forward the calls to Databricks H3 when it is available (DBR 11.2 + Photo, or DBR 11.2 and premium plan), but will fallback to the standard Mosaic/H3 implementation otherwise. So Mosaic will keep working without photon enabled. We still recommend Photon because it's faster, but it's not required.

**Comment by mickvanhulst-TomTom** [comment]
@edurdevic has Photon been tested with Mosaic? Would be great if you could do some performance comparisons :)!
